django/cantusdb_project/main_app/signals.py

```diff
@@ -28,32 +53,120 @@ def update_chant_search_vector(instance, **kwargs):
+        # the `searchMelody.js` on old cantus makes no reference to the b-flat accidentals ("y", "i", "z")
+        # so put them in unwanted chars for now
+        unwanted_chars = [
+            "-",
+            "1",
+            "2",
+            "3",
+            "4",
+            "5",
+            "6",
+            "7",
+            "?",
+            ".",
+            " ",
+            "y",
+            "i",
+            "z",
+        ]
+        # convert all charactors to lower-case, upper-case letters stand for liquescent of the same pitch
+        volpiano_lower = volpiano.lower()
+        # `)` stands for the lowest `g` note liquescent in volpiano, its 'lower case' is `9`
+        volpiano_notes = volpiano_lower.replace(")", "9")
+        # remove none-note charactors
+        for unwanted_char in unwanted_chars:
+            volpiano_notes = volpiano_notes.replace(unwanted_char, "")
+        # remove duplicate consecutive chars
+        volpiano_notes = re.sub(r"(.)\1+", r"\1", volpiano_notes)
+        return volpiano_notes
+
+    def generate_volpiano_intervals(volpiano_notes):
+        """
+        Populate the ``volpiano_intervals`` field of the ``Chant`` model
+
+        This field is used for melody search when searching for transpositions
+
+        Args:
+            volpiano_notes (str): The content of ``chant.volpiano_notes``,
+            populated by the ``generate_volpiano_notes`` function
+
+        Returns:
+            str: A str of digits, recording the intervals between adjacent notes
+        """
+        # replace '9' (the note G) with the char corresponding to (ASCII(a) - 1), because 'a' denotes the note A
+        volpiano_notes = volpiano_notes.replace("9", chr(ord("a") - 1))
+        # we model the interval between notes using the difference between the ASCII codes of corresponding letters
+        # the letter for the note B is "j" (106), note A is "h" (104), the letter "i" (105) is skipped
+        # move all notes above A down by one letter
+        volpiano_notes = list(volpiano_notes)
+        for j, note in enumerate(volpiano_notes):
+            if ord(note) >= 106:
+                volpiano_notes[j] = chr(ord(note) - 1)
+
+        # `intervals` records the difference between two adjacent notes.
+        # Note that intervals are encoded by counting the number of scale
+        # steps between adjacent notes: an ascending second is thus encoded
+        # as "1"; a descending third is encoded "-2", and so on.
+        intervals = []
+        for j in range(1, len(volpiano_notes)):
+            intervals.append(ord(volpiano_notes[j]) - ord(volpiano_notes[j - 1]))
+        # convert `intervals` to str
+        volpiano_intervals = "".join([str(interval) for interval in intervals])
+        return volpiano_intervals
+
+    if instance.volpiano is None:
+        return
+
+    volpiano_notes = generate_volpiano_notes(
+        instance.volpiano
+    )
+    Chant.objects.filter(id=instance.id).update(
```

Comment:
This should be `Chant.objects.get(id=instance.id)` because there would only be one chant object with this id. 
Now that I think about it, why not just do `instance.update()`?

Reply:
I copied `Chant.objects.filter().update()` from elsewhere in the codebase. It turns out that if you use `.get().update()` or just `instance.update()`, python raises `AttributeError: 'Chant' object has no attribute 'update'`.

That said, this is a good place for a simplification/refactoring - you can update multiple fields at once with a single `.update()`. I'll push an update that makes this change presently.